dashboard/app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -5,6 +5,60 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
 
   # GET /users/auth/:provider/callback
   def all
+    if should_connect_provider?
+      connect_provider
+    else
+      login
+    end
+  end
+
+  # GET /users/auth/:provider/connect
+  def connect
+    return head(:bad_request) unless current_user&.migrated? && params[:provider].present?
+
+    session[:connect_provider] = 2.minutes.from_now
+    redirect_to omniauth_authorize_path(current_user, params[:provider])
+  end
+
+  # Call GET /users/auth/:provider/connect and the callback will trigger this code path
+  def connect_provider
+    return head(:bad_request) unless can_connect_provider? # TODO: (madelynkasula) better error here?
+
+    auth_hash = request.env['omniauth.auth']
+    provider = auth_hash.provider.to_s
+    return head(:bad_request) unless AuthenticationOption::OAUTH_CREDENTIAL_TYPES.include? provider
+
+    # TODO: some of this won't work right for non-Google providers, because info comes in differently
+    new_data = nil
+    if auth_hash.credentials && (auth_hash.credentials.token || auth_hash.credentials.expires_at || auth_hash.credentials.refresh_token)
+      new_data = {
+        oauth_token: auth_hash.credentials.token,
+        oauth_token_expiration: auth_hash.credentials.expires_at,
+        oauth_refresh_token: auth_hash.credentials.refresh_token
+      }.to_json
+    end
+    email = auth_hash.info.email
+    hashed_email = nil
+    hashed_email = User.hash_email(email) unless email.blank?
+    auth_option = AuthenticationOption.new(
+      user: current_user,
+      email: email,
+      hashed_email: hashed_email || '',
+      credential_type: provider,
+      authentication_id: auth_hash.uid,
+      data: new_data
+    )
+
+    if auth_option.save
+      return head(:no_content)
+    else
+      render status: :unprocessable_entity,
+             json: auth_option.errors.as_json(full_messages: true),
+             content_type: 'application/json'
+    end
+  end
+
+  def login
```

Comment:
Note for future cleanup: `login` still contains some a non-login concern, the `just_authorized_google_classroom` path.  It would probably be cleaner to break this logic into its own method and make it part of the top-level decision in `all`.